### PR TITLE
feat(accont-lib): add deposit and stake builder for near

### PR DESCRIPTION
**Pull request by facundoCarreSimtlix**
add deposit and stake builder for near

STLX-14821

### CODE REVIEW

File: modules/account-lib/src/coin/near/contractCallTxBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import { BuildTransactionError } from '../baseCoin/errors';
+
+export interface functionCallParams {
+  methodName: string;
+  args: { amount: string };
+  gas: BN;
+  deposit: BN;
+}
+
+export abstract class ContractCallTxBuider extends TransactionBuilder {
```
Comment (ipelaez-simtlix) [inline modules/account-lib/src/coin/near/contractCallTxBuilder.ts:13]: to be consinstent with others builders, we should call it ContractCallBuilder.

File: modules/account-lib/src/coin/near/contractCallTxBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import * as nearAPI from 'near-api-js';
+import { BuildTransactionError } from '../baseCoin/errors';
+
+export interface functionCallParams {
+  methodName: string;
+  args: { amount: string };
+  gas: BN;
+  deposit: BN;
+}
+
```
Comment (ipelaez-simtlix) [inline modules/account-lib/src/coin/near/contractCallTxBuilder.ts:12]: can we move it to iface.ts? maybe on a staking #region

File: modules/account-lib/src/coin/near/contractCallTxBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    if (!this.isValidAmount(params.gas)) {
+      throw new BuildTransactionError(`Invalid gas amount ${params.gas}`);
+    }
+    this._gas = params.gas;
+    if (!this.isValidAmount(params.deposit)) {
+      throw new BuildTransactionError(`Invalid deposit amount ${params.deposit}`);
+    }
+    this._deposit = params.deposit;
+    return this;
+  }
```
Comment (ipelaez-simtlix) [inline modules/account-lib/src/coin/near/contractCallTxBuilder.ts:31]: lets change it, for inherited builders we already know which is the method name so we have two alternatives, do this builder totally agnostic and use an object like the current implementation, or make methods in order to let inherited builders set those params.

e.g
StakingActivateBuilder().amount('100').gas('100')
StakingDeactivateBuilder().amount('100').gas('100')
StakingWithdrawBuilder().amount('100').gas('100')

Is code responsability check if we need internally set the .amount() on the _deposit or the _methodName calling the ContractCallBuilder.
Is staking builder responsability set the method name ( we already know it )

File: modules/account-lib/src/coin/near/iface.ts [hunk 4]
```diff
@@ -19,9 +19,21 @@ export interface Transfer {
...
+}
+
+/**
+ * Actions implemented on Near account-lib
+ * add here as more are implemented.
+ */
 export interface Action {
-  // TODO: add actions when they are implemented
   transfer: Transfer;
+  functionCall: FunctionCall;
```
Comment (silalvado) [inline modules/account-lib/src/coin/near/iface.ts:36]: Shouldn't these fields be optional?

File: modules/account-lib/src/coin/near/transaction.ts [hunk 5]
```diff
@@ -126,30 +141,48 @@ export class Transaction extends BaseTransaction {
...
-      throw new InvalidTransactionError('unsupported action in raw transaction');
+    const action = this._nearTransaction.actions[0];
+
+    switch (action.enum) {
+      case 'transfer':
+        this.setTransactionType(TransactionType.Send);
+        break;
+      case 'functionCall':
+        const methodName = action.functionCall.methodName;
+        if (methodName in StakingContractMethodNames) {
```
Comment (silalvado) [inline modules/account-lib/src/coin/near/transaction.ts:172]: I think this is wrong. If the method is in the list of valid methods shouldn't throw an error. "In" is used with objects, I'm not sure that it works with enums.

File: modules/account-lib/src/coin/near/contractCallWrapper.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+  /** Set gas, expresed on yoctol */
+  public set gas(gas: string) {
+    if (!this.isValidAmount(new BigNumber(gas))) {
+      throw new InvalidParameterValueError('Invalid gas value');
+    }
+    this._gas = gas;
+  }
+
+  /** Get gas, expresed on yoctol */
```
Comment (silalvado) [inline modules/account-lib/src/coin/near/contractCallWrapper.ts:32]: yoctos

**Review by silalvado — CHANGES_REQUESTED** [review]

File: modules/account-lib/src/coin/near/transaction.ts (hunk 5 above)
Comment (ipelaez-simtlix) [reply modules/account-lib/src/coin/near/transaction.ts:172]: Its wrong, nice catch

Comment (ipelaez-simtlix) [reply modules/account-lib/src/coin/near/transaction.ts:172]: Its doing a fake falsy actually, making that it "works" but its a bug.

File: modules/account-lib/src/coin/near/contractCallWrapper.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    this._args = args;
+  }
+
+  /** Set args, which are the parameters of a method */
+  public get args(): Record<string, unknown> {
+    return this._args;
+  }
+
+  /**
+   * get all parameters of the contractCallWrapper
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/near/contractCallWrapper.ts:61]: nit: Get

File: modules/account-lib/src/coin/near/transaction.ts [hunk 8]
```diff
@@ -184,6 +240,25 @@ export class Transaction extends BaseTransaction {
     };
   }
 
+  /**
+   * Returns a complete explanation for a staking activate transaction
+   * @param {TxData} json The transaction data in json format
+   * @param {TransactionExplanation} explanationResult The transaction explanation to be completed
+   * @returns {TransactionExplanation}
+   */
+  explainStakingActivateTransaction(json: TxData, explanationResult: TransactionExplanation): TransactionExplanation {
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/near/transaction.ts:249]: I don't know if for this case we only use explain tx and do each explain using the tx type. We need to think about this in order to have splitted the explaintx method.  Check other coins.

File: modules/account-lib/src/coin/near/transaction.ts [hunk 9]
```diff
@@ -193,7 +268,7 @@ export class Transaction extends BaseTransaction {
       // txhash used to identify the transactions
       id: result.id || '',
       displayOrder,
-      outputAmount: result.actions[0].transfer.deposit.toString(),
+      outputAmount: '',
```
Comment (rliberal-simtlix) [inline modules/account-lib/src/coin/near/transaction.ts:271]: why output here is empty?

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+      tx.outputs.length.should.equal(1);
+      tx.outputs[0].should.deepEqual({
+        address: validatorContractAddress,
+        value: amount,
+        coin: 'tnear',
+      });
+
+      should.equal(tx.type, TransactionType.StakingActivate);
+      const rawTx = tx.toBroadcastFormat();
+      should.equal(rawTx, testData.rawTx.stakingActivate.signed);
```
Comment (silalvado) [inline modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:40]: I think you should check tx.toJson(), particularly actions.

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      txBuilder
+        .gas(gas)
+        .sender(testData.accounts.account1.address, testData.accounts.account1.publicKey)
+        .receiverId(validatorContractAddress)
+        .recentBlockHash(testData.blockHash.block1)
+        .nonce(1);
+      await txBuilder.build().should.be.rejectedWith('amount is required before building staking activate');
+    });
+  });
+});
```
Comment (silalvado) [inline modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:93]: You should also test fromImplementation for staking transactions.

**Review by silalvado — CHANGES_REQUESTED** [review]

File: modules/account-lib/src/coin/near/iface.ts (hunk 4 above)
Comment (ipelaez-simtlix) [reply modules/account-lib/src/coin/near/iface.ts]: we resolved this offline

File: modules/account-lib/src/coin/near/transaction.ts (hunk 9 above)
Comment (ipelaez-simtlix) [reply modules/account-lib/src/coin/near/transaction.ts:272]: hmm so based on other coins, it should be initially zero, right?

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts (hunk 10 above)
Comment (rliberal-simtlix) [reply modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:40]: good catch!

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts (hunk 11 above)
Comment (rliberal-simtlix) [reply modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:93]: good catch!

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        coin: 'tnear',
+      });
+
+      should.equal(tx.type, TransactionType.StakingActivate);
+      const rawTx = tx.toBroadcastFormat();
+      should.equal(rawTx, testData.rawTx.stakingActivate.signed);
+      const txJson = tx.toJson();
+      txJson.should.have.properties(['id', 'signerId', 'publicKey', 'nonce', 'actions', 'signature']);
+      txJson.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
+      txJson.signerId.should.equal('61b18c6dc02ddcabdeac56cb4f21a971cc41cc97640f6f85b073480008c53a0d');
```
Comment (silalvado) [inline modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:44]:
```suggestion
      txJson.signerId.should.equal('testData.accounts.account1.address');
```

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        .nonce(1);
+      await txBuilder.build().should.be.rejectedWith('amount is required before building staking activate');
+    });
+
+    it('build from an unsigned staking activate', async () => {
+      const txBuilder = factory.from(testData.rawTx.stakingActivate.unsigned);
+      const tx = await txBuilder.build();
+      tx.inputs.length.should.equal(1);
+      tx.inputs[0].should.deepEqual({
+        address: '61b18c6dc02ddcabdeac56cb4f21a971cc41cc97640f6f85b073480008c53a0d',
```
Comment (silalvado) [inline modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:115]:
```suggestion
        address: testData.accounts.account1.address,
```

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+      });
+
+      should.equal(tx.type, TransactionType.StakingActivate);
+      const rawTx = tx.toBroadcastFormat();
+      should.equal(rawTx, testData.rawTx.stakingActivate.signed);
+      const txJson = tx.toJson();
+      txJson.should.have.properties(['id', 'signerId', 'publicKey', 'nonce', 'actions', 'signature']);
+      txJson.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
+      txJson.signerId.should.equal('61b18c6dc02ddcabdeac56cb4f21a971cc41cc97640f6f85b073480008c53a0d');
+      txJson.publicKey.should.equal('ed25519:7aMa9bTKziRzGACadUGwSLi6BRjtQQzHJPyzQrpZvj4G');
```
Comment (silalvado) [inline modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts:45]: Maybe we can add this key to the near resources

**Review by silalvado — CHANGES_REQUESTED** [review]

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts (hunk 14 above)
Comment (ipelaez-simtlix) [reply modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts]: done

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+      });
+      tx.id.should.equal('GpiLLaGs2Fk2bd7SQvhkJaZjj74UnPPdF7cUa9pw15je');
+    });
+
+    it('build from an signed staking activate', async () => {
+      const txBuilder = factory.from(testData.rawTx.stakingActivate.signed);
+      const tx = await txBuilder.build();
+      tx.inputs.length.should.equal(1);
+      tx.inputs[0].should.deepEqual({
+        address: '61b18c6dc02ddcabdeac56cb4f21a971cc41cc97640f6f85b073480008c53a0d',
```
[the remark this answers is not in the document]
Comment (ipelaez-simtlix) [reply modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts]: done

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts (hunk 13 above)
Comment (ipelaez-simtlix) [reply modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts]: done

File: modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts (hunk 12 above)
Comment (ipelaez-simtlix) [reply modules/account-lib/test/unit/coin/near/transactionBuilder/stakingActivateBuilder.ts]: done

**Review by ewangbitgo — APPROVED** [review]
lgtm

**Review by rliberal-simtlix — APPROVED** [review]

**Review by silalvado — APPROVED** [review]

**Pull request merged**